File: sdk/ios/Sources/FrakSDKUI/SharingWebView.swift
#if canImport(UIKit)
    import SwiftUI
    import WebKit

    /// One sheet's worth of wiring for a `SharingWebView`. Split from the view so a pooled view
    /// can outlive a sheet; rebinding also resets the view's per-load state.
    struct SharingWebViewBinding {
        let sessionId: String
        let onAction: (SharingPageAction) -> Void
        let onPageReady: () -> Void
        let onLoadFailed: () -> Void
        let onOpenExternal: (URL) -> Void

        init(
            sessionId: String,
            onAction: @escaping (SharingPageAction) -> Void = { _ in },
            onPageReady: @escaping () -> Void = {},
            onLoadFailed: @escaping () -> Void = {},
            onOpenExternal: @escaping (URL) -> Void = { _ in }
        ) {
            self.sessionId = sessionId
            self.onAction = onAction
            self.onPageReady = onPageReady
            self.onLoadFailed = onLoadFailed
            self.onOpenExternal = onOpenExternal
        }

        /// What an unpresented, warm view carries: `warmSessionId` can never satisfy a real
        /// sheet's `sid` guard. Computed, since the callbacks make this type non-`Sendable`.
        static var warm: SharingWebViewBinding { SharingWebViewBinding(sessionId: SharingPageURL.warmSessionId) }
    }

    /// The web view the sheet loads the hosted page in, and the navigation policy that is its
    /// only channel back to the host. No JavaScript bridge: the page reports by navigating to
    /// `returnScheme://result?sid=…&action=…`, intercepted here and never allowed to leave the app.
    @MainActor
    final class SharingWebView: NSObject {
        let view: WKWebView

        private let origin: URL?
        private let returnScheme: String

        private(set) var binding: SharingWebViewBinding

        /// The document `load(_:)` last pointed at, minus any fragment. Tracked here rather than
        /// read from `WKWebView.url`, which reports the committed URL and so lags a load in flight.
        private(set) var loadedBaseURL: String?

        /// Whether `loadedBaseURL` finished loading. A fragment change starts no request, so
        /// activating on top of a half-loaded document would leave the page stuck where it got to.
        private(set) var documentReady = false

        /// Jetsammed content process, not loaded since. WebKit gives the view a new process on the
        /// next load, so this only tells `SharingWebViewPool` its warm URL is stale.
        private(set) var rendererGone = false

        /// The last main-frame URL asked for, so a retry has something to retry.
        private var requested: URL?
        /// Rungs of `retryLadder` already spent on the document in `ladderURL`.
        private var retryCount = 0
        /// Which document the spent rungs belong to. A session navigates more than once — the
        /// install page, and the confirmation screen — and a fresh document has not failed yet.
        private var ladderURL: URL?
        /// The scheduled retry, held so a rebind can cancel one that would navigate the next
        /// session's view.
        private var pendingRetry: Task<Void, Never>?
        /// Set between issuing the retry and it starting, so the original load's duplicate
        /// failure callbacks are not read as the retry failing too.
        private var retryPending = false
        /// `onLoadFailed` is called at most once per binding.
        private var settled = false
        /// Set when the in-flight main-frame navigation fails, so the `didFinish` WebKit still
        /// delivers for the error document is not read as a load.
        private var navigationFailed = false

        /// Whether the navigation in flight was started under the current binding. A pooled view
        /// is bound mid-flight, so the warm load's `didFinish` or failure must not reach the sheet.
        private var navigationOwnedByBinding = false

        init(
            walletOrigin: String,
            returnScheme: String,
            binding: SharingWebViewBinding = .warm
        ) {
            let configuration = WKWebViewConfiguration()
            // Persistent: the hosted page's own HTTP cache is what tier 2 falls back on.
            configuration.websiteDataStore = .default()
            configuration.preferences.javaScriptCanOpenWindowsAutomatically = false

            // Screen-sized, not `.zero`: a warm view is never in a window, and at 0x0 the page's
            // whole first layout — every `innerWidth`, media query and container query — is made
            // against a degenerate viewport it then has to recover from when the sheet shows it.
            self.view = WKWebView(frame: Self.warmFrame(), configuration: configuration)
            self.origin = URL(string: walletOrigin)
            self.returnScheme = returnScheme
            self.binding = binding
            super.init()

            view.allowsLinkPreview = false
            view.allowsBackForwardNavigationGestures = false
            view.isOpaque = false
            view.backgroundColor = .clear
            view.navigationDelegate = self
            #if DEBUG
                if #available(iOS 16.4, *) {
                    // The hosted page is the one part of the sheet that can fail silently.
                    view.isInspectable = true
                }
            #endif

            // The view fills the sheet, home indicator included, and the page insets its own
            // footer from `env(safe-area-inset-bottom)`. Any other behaviour insets the document
            // by that same safe area, leaving the sheet showing through under the page.
            view.scrollView.contentInsetAdjustmentBehavior = .never
            // The document never scrolls — the page scrolls a child of its own — so a bounce here
            // is only a rubber-band competing with the sheet's drag.
            view.scrollView.bounces = false
        }

        /// A plausible viewport for a view that is not in a window yet. The constant is a
        /// mid-range iPhone; only its non-degeneracy matters, since the sheet resizes the view.
        private static func warmFrame() -> CGRect {
            let scene = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }.first
            return scene?.screen.bounds ?? CGRect(x: 0, y: 0, width: 390, height: 844)
        }

        /// Points the view at a session. Resets per-load state; see `SharingWebViewBinding`.
        func bind(_ binding: SharingWebViewBinding) {
            self.binding = binding
            // Before the counters: a retry booked by the previous session would otherwise navigate
            // this view to that session's URL, on a pool that has already moved on.
            cancelPendingRetry()
            requested = nil
            retryCount = 0
            ladderURL = nil
            retryPending = false
            settled = false
            navigationFailed = false
            navigationOwnedByBinding = false
        }

        /// A full navigation; fragment activations go through `navigate(_:)` instead.
        ///
        /// - Parameters:
        ///   - url: the address to load.
        ///   - baseURL: the caller's own string, compared with `==` against
        ///     `SharingSession.warmBaseURL`, so it must not rely on `URL(string:)` round-tripping.
        func load(_ url: URL, baseURL: String? = nil) {
            loadedBaseURL = (baseURL ?? url.absoluteString).components(separatedBy: "#")[0]
            documentReady = false
            // This load is what gives a jetsammed view its new content process back.
            rendererGone = false
            requested = url
            view.load(URLRequest(url: url))
        }

        func onDocumentReady() {
            documentReady = true
        }

        /// Performs a `SharingNavigation`. The activation case hangs its fragment off
        /// `WKWebView.url` rather than off the URL we warmed with, so it stays correct if the two
        /// ever diverge. A fragment-only load fires no `didFinish`.
        func navigate(_ navigation: SharingNavigation) {
            switch navigation {
            case .load(let url):
                load(url)
            case .activate(let fragment, let fullURL):
                guard let target = fragmentTarget(fragment) else {
                    // No committed URL means no document to hang a fragment off; load the page.
                    load(fullURL)
                    return
                }
                // Not through `load(_:)`: same document, so `loadedBaseURL`, `documentReady` and
                // `requested` must keep describing it.
                view.load(URLRequest(url: target))
            }
        }

        /// Puts this view's page back to its warm params after a session moved them. Same document,
        /// so `loadedBaseURL` and `documentReady` keep describing it and the next session can
        /// activate on top instead of loading.
        func resetToWarm() {
            guard let target = fragmentTarget(SharingPageURL.warmFragment) else { return }
            view.load(URLRequest(url: target))
        }

        /// `fragment` hung off whatever document is committed, or nil when there is none.
        private func fragmentTarget(_ fragment: String) -> URL? {
            guard let committed = view.url?.absoluteString.components(separatedBy: "#")[0] else {
                return nil
            }
            return URL(string: committed + fragment)
        }

        func stopLoading() {
            view.stopLoading()
        }

        /// Retires the view for good: the delegate is dropped, so nothing can reach a binding after.
        func destroy() {
            // The pool reaches here without rebinding — a dead pool releasing a lent view, or
            // destroying a warm one — so the binding's own cancellation does not cover it, and a
            // booked retry would keep the view alive to load a page nobody will see.
            cancelPendingRetry()
            view.navigationDelegate = nil
            view.stopLoading()
        }

        /// Component by component, never a prefix match: `wallet.frak.id.attacker.example`
        /// starts with the origin string but is not it.
        private func isSameOrigin(_ url: URL) -> Bool {
            guard let origin else { return false }
            return url.scheme == origin.scheme
                && url.host?.caseInsensitiveCompare(origin.host ?? "") == .orderedSame
                && url.port == origin.port
        }

        private func queryValue(_ url: URL, _ name: String) -> String? {
            sharingQueryValue(url, name)
        }

        /// A main-frame failure gets the rest of `retryLadder` before tier 3. Every rung is a
        /// network rung: the hosted document is served `no-store`, so it is never in the HTTP
        /// cache and a cache-only attempt cannot answer.
        ///
        /// - Parameter unreachable: the network itself did not answer, so another attempt over it
        ///   is pointless and there is no cached copy to read instead — tier 3 now.
        private func handleMainFrameFailure(unreachable: Bool) {
            // The warm load failing after this view was lent to a sheet: not the sheet's page.
            guard navigationOwnedByBinding else { return }
            // Before the `settled` guard: a later error document's `didFinish` must not report readiness.
            navigationFailed = true
            guard !settled else { return }
            // Duplicate callback for the failure that already booked the next rung.
            guard !retryPending else { return }
            guard let requested else {
                giveUp()
                return
            }
            // A document this ladder has not been spent on yet gets the whole thing.
            if requested != ladderURL {
                ladderURL = requested
                retryCount = 0
            }
            guard !unreachable, retryCount < Self.retryLadder.count else {
                giveUp()
                return
            }
            let delay = Self.retryLadder[retryCount]
            retryCount += 1
            retryPending = true
            scheduleRetry(after: delay) { [weak self] in
                guard let self else { return }
                view.load(URLRequest(url: requested, cachePolicy: .useProtocolCachePolicy))
            }
        }

        /// The ladder is spent.
        private func giveUp() {
            settled = true
            binding.onLoadFailed()
        }

        private func scheduleRetry(
            after delay: TimeInterval,
            _ navigate: @escaping @MainActor () -> Void
        ) {
            cancelPendingRetry()
            pendingRetry = Task { @MainActor [weak self] in
                if delay > 0 {
                    try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                }
                // Outside the delay branch on purpose: `Task.cancel()` is cooperative, and the
                // undelayed rung is the one most likely to race a rebind. A `Task` does not run
                // inline, so a synchronous `bind()` on the same turn can cancel this before it
                // starts — and without this check it would navigate a view the pool reassigned.
                guard !Task.isCancelled else { return }
                self?.pendingRetry = nil
                self?.retryPending = false
                navigate()
            }
        }

        private func cancelPendingRetry() {
            pendingRetry?.cancel()
            pendingRetry = nil
        }

        /// The network itself did not answer, so another attempt over it is pointless.
        private func isUnreachable(_ error: any Error) -> Bool {
            let error = error as NSError
            guard error.domain == NSURLErrorDomain else { return false }
            return error.code == NSURLErrorNotConnectedToInternet
                || error.code == NSURLErrorCannotFindHost
                || error.code == NSURLErrorCannotConnectToHost
                || error.code == NSURLErrorDNSLookupFailed
        }

        /// Delays a main-frame failure gets before tier 3, in order. Two rungs, sized to fit
        /// inside the sheet's own load budget alongside the attempts themselves.
        private static let retryLadder: [TimeInterval] = [0.3, 0.9]

        /// A navigation this code cancelled, which WebKit reports as a load failure.
        private func isCancellation(_ error: any Error) -> Bool {
            let error = error as NSError
            if error.domain == NSURLErrorDomain, error.code == NSURLErrorCancelled { return true }
            // `WebKitErrorFrameLoadInterruptedByPolicyChange`: legacy domain and code, no symbol
            // in `WKError`. What a `.cancel` decision surfaces as.
            return error.domain == "WebKitErrorDomain" && error.code == 102
        }
    }

    extension SharingWebView: WKNavigationDelegate {
        func webView(
            _ webView: WKWebView,
            decidePolicyFor navigationAction: WKNavigationAction,
            decisionHandler: @escaping @MainActor @Sendable (WKNavigationActionPolicy) -> Void
        ) {
            guard let url = navigationAction.request.url else {
                decisionHandler(.cancel)
                return
            }

            // Sub-frames are never launched externally, and a cross-origin one is cancelled rather
            // than rendered. Only remote schemes are judged — `about:`/`blob:`/`data:` have no host.
            if let frame = navigationAction.targetFrame, !frame.isMainFrame {
                let remote = url.scheme == "https" || url.scheme == "http"
                // The return scheme carries a capability value (`action=code`); cancel it outright,
                // never merely leave it unhandled.
                if url.scheme == returnScheme {
                    decisionHandler(.cancel)
                    return
                }
                decisionHandler(remote && !isSameOrigin(url) ? .cancel : .allow)
                return
            }

            // A nil `targetFrame` is a new window (`target="_blank"`, `window.open`), which
            // `javaScriptCanOpenWindowsAutomatically = false` does not stop and `.allow` would drop.
            if navigationAction.targetFrame == nil, isSameOrigin(url) {
                webView.load(navigationAction.request)
                decisionHandler(.cancel)
                return
            }

            if url.scheme == returnScheme, url.host == SharingPageURL.resultHost {
                // A stale sid: a closed sheet's result, or the warm page's, which no sheet holds.
                if queryValue(url, "sid") == binding.sessionId,
                    let name = queryValue(url, "action"),
                    let action = SharingPageAction.from(
                        action: name,
                        value: queryValue(url, "value"),
                        exp: queryValue(url, "exp"),
                        shareTitle: queryValue(url, "title"),
                        shareText: queryValue(url, "text"),
                        shareImage: queryValue(url, "image")
                    )
                {
                    binding.onAction(action)
                }
                decisionHandler(.cancel)
                return
            }

            if isSameOrigin(url) {
                decisionHandler(.allow)
                return
            }

            // Anything else belongs in the browser, not inside a sheet the user cannot navigate.
            binding.onOpenExternal(url)
            decisionHandler(.cancel)
        }

        func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
            retryPending = false
            navigationFailed = false
            navigationOwnedByBinding = true
        }

        /// Without this the main-frame status code is never inspected, and a 5xx that returns a
        /// body reaches `didFinish` as a normal load.
        func webView(
            _ webView: WKWebView,
            decidePolicyFor navigationResponse: WKNavigationResponse,
            decisionHandler: @escaping @MainActor @Sendable (WKNavigationResponsePolicy) -> Void
        ) {
            guard navigationResponse.isForMainFrame,
                let http = navigationResponse.response as? HTTPURLResponse,
                !(200..<400).contains(http.statusCode)
            else {
                decisionHandler(.allow)
                return
            }
            // `.allow`, not `.cancel`: cancelling surfaces as a cancellation error that
            // `isCancellation` filters out. `navigationFailed` suppresses the `didFinish`.
            decisionHandler(.allow)
            // An answer, however bad, means the network is there; this is the retryable kind.
            handleMainFrameFailure(unreachable: false)
        }

        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            // The warm load landing after this view was lent to a sheet. Not this session's page.
            guard navigationOwnedByBinding else { return }
            guard !navigationFailed else { return }
            binding.onPageReady()
        }

        func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: any Error) {
            guard !isCancellation(error) else { return }
            handleMainFrameFailure(unreachable: isUnreachable(error))
        }

        func webView(
            _ webView: WKWebView,
            didFailProvisionalNavigation navigation: WKNavigation!,
            withError error: any Error
        ) {
            guard !isCancellation(error) else { return }
            handleMainFrameFailure(unreachable: isUnreachable(error))
        }

        /// A jetsammed content process leaves a blank view and fires nothing else.
        func webViewWebContentProcessDidTerminate(_ webView: WKWebView) {
            // Ahead of the `settled` guard: a jetsammed renderer leaves nothing on screen, so a
            // pooled view still claiming `documentReady` would activate into a dead renderer.
            documentReady = false
            // Tells the pool its warm URL describes nothing on screen.
            rendererGone = true
            // A booked retry would only navigate a view with nothing behind it.
            cancelPendingRetry()
            guard !settled else { return }
            settled = true
            binding.onLoadFailed()
        }
    }

    /// Puts an already-built `SharingWebView` on screen; the sheet's model owns it, not SwiftUI.
    struct SharingWebViewContainer: UIViewRepresentable {
        let webView: SharingWebView
        /// Called once SwiftUI has actually taken this view out of the hierarchy — not from
        /// `onDismiss`, which still has frames to draw and would empty the closing sheet.
        let onDismantled: () -> Void

        func makeCoordinator() -> Coordinator {
            Coordinator(onDismantled: onDismantled)
        }

        /// The engine itself, not a host wrapping it: SwiftUI builds this representable twice when
        /// `SharingPresenter.presentation` swaps, and a host would have to *move* the engine into
        /// the second one, leaving the on-screen first one empty.
        func makeUIView(context: Context) -> WKWebView {
            return webView.view
        }

        func updateUIView(_ uiView: WKWebView, context: Context) {}

        static func dismantleUIView(_ uiView: WKWebView, coordinator: Coordinator) {
            coordinator.onDismantled()
        }

        /// Carries the callback, which `dismantleUIView` cannot reach any other way — it is
        /// static, and the representable value is gone by the time it runs.
        final class Coordinator {
            let onDismantled: () -> Void

            init(onDismantled: @escaping () -> Void) {
                self.onDismantled = onDismantled
            }
        }
    }
#endif

File: sdk/android/frak-sdk-ui/src/main/kotlin/id/frak/sdk/ui/SharingWebView.kt
package id.frak.sdk.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.ApplicationInfo
import android.graphics.Bitmap
import android.graphics.Color
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.webkit.ConsoleMessage
import android.webkit.CookieManager
import android.webkit.RenderProcessGoneDetail
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient

/** What the hosted page told the host. Mirrors the wallet's `HostResultAction`. */
internal sealed interface SharingPageAction {
    data object Install : SharingPageAction

    data object Dismiss : SharingPageAction

    data object ShareAgain : SharingPageAction

    /** The page asking the host to share: navigator.share does not exist in an Android WebView. */
    data object Share : SharingPageAction

    data object Copy : SharingPageAction

    data object Error : SharingPageAction

    /** The page reporting that it has painted; [SharingWebViewBinding.onPageVisible] is the fallback. */
    data object Ready : SharingPageAction

    data class Code(
        val value: String,
        /** Epoch seconds, or null when the page sent none. */
        val expiresAtSeconds: Long?,
    ) : SharingPageAction

    companion object {
        /** Unknown actions are null: the page can ship a new one before the SDK that reads it. */
        fun fromWire(
            action: String?,
            value: String? = null,
            exp: String? = null,
        ): SharingPageAction? =
            when (action) {
                "install" -> Install

                "dismiss" -> Dismiss

                "shareAgain" -> ShareAgain

                "share" -> Share

                "copy" -> Copy

                "error" -> Error

                "ready" -> Ready

                // A code action with no code is not one; treat it as unknown.
                "code" -> value?.takeIf { it.isNotEmpty() }?.let { Code(it, exp?.toLongOrNull()) }

                else -> null
            }
    }
}

/** One sheet's worth of wiring for a [SharingWebViewHandle]. Rebinding resets per-load state. */
internal class SharingWebViewBinding(
    val sessionId: String,
    val onAction: (SharingPageAction) -> Unit = {},
    val onPageReady: () -> Unit = {},
    /** First paint, as opposed to [onPageReady]'s document-finished. Drives the skeleton. */
    val onPageVisible: () -> Unit = {},
    val onLoadFailed: () -> Unit = {},
    val onOpenExternal: (String) -> Unit = {},
) {
    companion object {
        val Warm: SharingWebViewBinding = SharingWebViewBinding(sessionId = WARM_SESSION_ID)

        /** Never a real sheet's id, so a warm page's result navigation is never attributed to one. */
        const val WARM_SESSION_ID: String = "warm"
    }
}

/** A sharing web view and its client, held here rather than read off it: `getWebViewClient` is API 26. */
internal class SharingWebViewHandle(
    val view: WebView,
    private val client: SharingWebViewClient,
) {
    /** The document [load] last pointed at; `WebView.getUrl` lags for the whole of a load. */
    var loadedBaseUrl: String? = null
        private set

    /** A fragment hung off a half-loaded document never finishes: no request is started. */
    var documentReady: Boolean = false
        private set

    /**
     * This view's render process was reclaimed or crashed. Terminal: Android's contract is that a
     * `WebView` whose renderer is gone is unusable for good, so [SharingWebViewPool] discards it
     * rather than loading into it again.
     */
    var rendererGone: Boolean = false
        private set

    /**
     * Called from [SharingWebViewClient.onRenderProcessGone], which cannot reach this handle
     * itself. Clearing [documentReady] is what stops the next sheet activating by fragment into a
     * document that no longer exists.
     */
    fun onRendererGone() {
        rendererGone = true
        documentReady = false
    }

    /** Points the view at a session. Resets per-load state; see [SharingWebViewBinding]. */
    fun bind(binding: SharingWebViewBinding) {
        // Belt and braces: nothing sets a non-default cache mode any more, and a pinned one would
        // be inherited silently by the pool's re-warm and by every later sheet.
        view.settings.cacheMode = WebSettings.LOAD_DEFAULT
        client.binding = binding
    }

    /** A full navigation. Fragment activations do not come through here — see [documentReady]. */
    fun load(url: String) {
        loadedBaseUrl = url.substringBefore('#')
        documentReady = false
        view.loadUrl(url)
    }

    fun onDocumentReady() {
        documentReady = true
    }

    var paused: Boolean = false
        private set

    /**
     * Background state while nobody is looking. `onPause` does not pause JavaScript; `pauseTimers`
     * would, but it is process-global and would reach the merchant's own web views.
     */
    fun pause() {
        if (paused) return
        paused = true
        view.onPause()
    }

    fun resume() {
        if (!paused) return
        paused = false
        view.onResume()
    }

    fun destroy() {
        // The client can be holding a retry scheduled against this view. It is posted to the main
        // looper, not to the view's own queue, so nothing else stops it — and `loadUrl` after
        // `destroy()` takes the merchant's process down. The pool reaches here without rebinding
        // first (a dead pool releasing a lent view, or destroying a warm one), so the binding
        // setter's own cancellation does not cover this.
        client.cancelPendingRetry()
        view.destroy()
    }
}

/**
 * Builds the sheet's WebView. There is no visible URL bar, hence the origin pinning and no JS
 * bridge. Returns unbound: nothing is reported until [SharingWebViewHandle.bind] names a session.
 */
@SuppressLint("SetJavaScriptEnabled")
internal fun createSharingWebView(
    context: Context,
    walletOrigin: String,
    returnScheme: String,
): SharingWebViewHandle {
    val client = SharingWebViewClient(walletOrigin = walletOrigin, returnScheme = returnScheme)
    // Gated on the host app's own debuggable flag, not on a Frak log level: `setWebContentsDebuggingEnabled`
    // is process-global and exposes the wallet session to anything that can reach the ADB socket.
    val debuggable = (context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
    if (debuggable) WebView.setWebContentsDebuggingEnabled(true)
    val view =
        WebView(context).apply {
            // Explicit MATCH_PARENT: a wrap-content WebView reports a 0 viewport height to Blink,
            // collapsing the page's `height: 100dvh` scroll container.
            layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, MATCH_PARENT)

            settings.javaScriptEnabled = true // Page is a React app; rest of this bounds what it can reach.
            settings.domStorageEnabled = true
            settings.allowFileAccess = false
            settings.allowContentAccess = false
            settings.mixedContentMode = WebSettings.MIXED_CONTENT_NEVER_ALLOW
            settings.setSupportMultipleWindows(false)
            settings.javaScriptCanOpenWindowsAutomatically = false
            settings.setGeolocationEnabled(false)
            settings.cacheMode = WebSettings.LOAD_DEFAULT // Explicit default: revalidate cached responses.

            // NORMAL, not the inherited NARROW_COLUMNS default, which reflows a responsive page.
            settings.layoutAlgorithm = WebSettings.LayoutAlgorithm.NORMAL

            // Transparent, not the default opaque white: the page rounds its own top corners, and
            // they can only cut through to the scrim if this view paints no background behind them.
            setBackgroundColor(Color.TRANSPARENT)

            // Third-party cookies off. First-party wallet cookies outlive the sheet regardless:
            // Android has no per-WebView data store, and the clear API is app-global.
            CookieManager.getInstance().setAcceptThirdPartyCookies(this, false)

            webViewClient = client
            // The page is the only thing in the sheet that can fail silently; without this a
            // merchant debugging a blank sheet has no signal at all.
            if (debuggable) webChromeClient = SharingConsoleClient()
        }

    // Registered on the view, not per load, so every wallet-origin document it shows is styled.
    // Result unused: the page's own CSS fallbacks are the degraded rendering.
    SharingHostStyle.install(view = view, walletOrigin = walletOrigin, topRadiusDp = SHEET_CORNER_RADIUS_DP)

    // Wired after construction rather than passed in: the client is built first, and this is the
    // only way back from it to the state a renderer crash invalidates.
    return SharingWebViewHandle(view = view, client = client).also {
        client.onRendererGone = it::onRendererGone
    }
}

/** Console forwarder for a debuggable host build. Returns true so the default logcat line is not also emitted. */
private class SharingConsoleClient : WebChromeClient() {
    override fun onConsoleMessage(message: ConsoleMessage): Boolean {
        Log.d(
            "FrakSharing",
            "page: ${message.message()} (${message.sourceId()}:${message.lineNumber()})",
        )
        return true
    }
}

internal class SharingWebViewClient(
    private val walletOrigin: String,
    private val returnScheme: String,
) : WebViewClient() {
    private val origin: Uri = Uri.parse(walletOrigin)

    /** Where a backed-off retry is posted. See [scheduleRetry] for why it is not the view's own queue. */
    private val handler = Handler(Looper.getMainLooper())

    /**
     * Told when this view's render process dies. Set once per view, not per binding: a renderer
     * crash outlives whichever session was using it, and the [SharingWebViewHandle] holding the
     * state it invalidates is not reachable from here.
     */
    var onRendererGone: () -> Unit = {}

    /** Whose session this view is currently serving. Rebinding clears every field below. */
    var binding: SharingWebViewBinding = SharingWebViewBinding.Warm
        set(value) {
            field = value
            // Before the counters: a retry booked by the previous session would otherwise navigate
            // this view to that session's URL, on a pool that has already moved on.
            cancelPendingRetry()
            pendingMainFrameUrl = null
            retryCount = 0
            ladderUrl = null
            retryPending = false
            settled = false
            navigationFailed = false
            navigationOwnedByBinding = false
        }

    /** A pooled view is bound mid-flight, and the warm load's callbacks must not count as the session's. */
    private var navigationOwnedByBinding = false

    /** Main-frame URL currently in flight, captured in [onPageStarted]; null once resolved. */
    private var pendingMainFrameUrl: String? = null

    /** Rungs of [RETRY_LADDER] already spent on the document in [ladderUrl]. */
    private var retryCount = 0

    /**
     * Which document the spent rungs belong to. A session navigates more than once — the install
     * page, and the confirmation screen — and a fresh document has not failed yet, so it must not
     * inherit a budget the sharing page spent.
     */
    private var ladderUrl: String? = null

    /** The scheduled retry, held so a rebind can cancel one that would navigate the next session's view. */
    private var pendingRetry: Runnable? = null

    /** So a duplicate error callback for the same failure is not read as the retry itself failing. */
    private var retryPending = false

    /** Set once [SharingWebViewBinding.onLoadFailed] fires, so it fires at most once per binding. */
    private var settled = false

    /** Android still fires [onPageFinished] for its own error page, which would read as success. */
    private var navigationFailed = false

    /** Distinguishes one [WebView.postVisualStateCallback] request from a stale earlier one. */
    private var visualStateRequest = 0L

    override fun shouldOverrideUrlLoading(
        view: WebView,
        request: WebResourceRequest,
    ): Boolean {
        // A sub-frame is never launched externally, and a cross-origin one is cancelled. Only
        // remote schemes are checked: about:blank/blob:/data: frames have no host to compare.
        // setSupportMultipleWindows(false) folds `target="_blank"` into the main frame, so it is never a sub-frame.
        if (!request.isForMainFrame) {
            val remote = request.url.scheme == "https" || request.url.scheme == "http"
            return remote && !isSameOrigin(request.url)
        }

        val url = request.url

        if (url.scheme == returnScheme && url.host == SharingPageUrl.RESULT_HOST) {
            // `sid` guards against a result from an already-closed sheet, or from the warm page.
            if (url.getQueryParameter("sid") == binding.sessionId) {
                SharingPageAction
                    .fromWire(
                        action = url.getQueryParameter("action"),
                        value = url.getQueryParameter("value"),
                        exp = url.getQueryParameter("exp"),
                    )?.let(binding.onAction)
            }
            return true
        }

        // Component-by-component compare: a prefix match would accept
        // `https://wallet.frak.id.attacker.example/`.
        if (isSameOrigin(url)) return false
        binding.onOpenExternal(url.toString())
        return true
    }

    private fun isSameOrigin(url: Uri): Boolean =
        url.scheme == origin.scheme &&
            url.host.equals(origin.host, ignoreCase = true) &&
            url.port == origin.port

    override fun onPageStarted(
        view: WebView,
        url: String,
        favicon: Bitmap?,
    ) {
        pendingMainFrameUrl = url
        retryPending = false
        navigationFailed = false
        navigationOwnedByBinding = true
    }

    override fun onPageFinished(
        view: WebView,
        url: String,
    ) {
        // The warm load landing after this view was lent to a sheet.
        if (!navigationOwnedByBinding) return
        // Android delivers this for its own error page too, in the same load cycle.
        if (navigationFailed) return
        // The ladder is NOT reset here: rungs belong to a document, not to an attempt at it, so a
        // load that only succeeded on its retry must not hand a later failure a full budget again.
        pendingMainFrameUrl = null
        val settledBinding = binding
        settledBinding.onPageReady()

        // Document-finished is not painted: React renders after `load`.
        val request = ++visualStateRequest
        view.postVisualStateCallback(
            request,
            object : WebView.VisualStateCallback() {
                override fun onComplete(requestId: Long) {
                    // A newer navigation superseded this one; its own callback owns visibility.
                    if (requestId != visualStateRequest) return
                    if (settledBinding !== binding) return
                    settledBinding.onPageVisible()
                }
            },
        )
    }

    override fun onReceivedError(
        view: WebView,
        request: WebResourceRequest,
        error: WebResourceError,
    ) {
        // Sub-resource failures degrade on their own; only the document matters.
        if (request.isForMainFrame) handleMainFrameFailure(view, unreachable = isUnreachable(error))
    }

    override fun onReceivedHttpError(
        view: WebView,
        request: WebResourceRequest,
        errorResponse: WebResourceResponse,
    ) {
        // An answer, however bad, means the network is there; this is the retryable kind.
        if (request.isForMainFrame) handleMainFrameFailure(view, unreachable = false)
    }

    /**
     * The network itself did not answer. Another attempt over it is pointless, so the ladder skips
     * straight to its cache-only rung rather than spending the sheet's budget on a dead radio.
     */
    private fun isUnreachable(error: WebResourceError): Boolean =
        error.errorCode == WebViewClient.ERROR_HOST_LOOKUP || error.errorCode == WebViewClient.ERROR_CONNECT

    /**
     * A main-frame failure gets the rest of [RETRY_LADDER] before tier 3. All rungs are network
     * rungs: the hosted document is served `no-store`, so it is never in the HTTP cache and a
     * cache-only attempt cannot answer — an unreachable network goes straight to tier 3.
     */
    private fun handleMainFrameFailure(
        view: WebView,
        unreachable: Boolean,
    ) {
        // The warm load failing after this view was lent to a sheet.
        if (!navigationOwnedByBinding) return
        // Before the `settled` guard: a later error page's onPageFinished must not report readiness.
        navigationFailed = true
        if (settled) return
        // Duplicate callback for the failure that already booked the next rung.
        if (retryPending) return
        val url = pendingMainFrameUrl
        if (url == null) {
            giveUp(view)
            return
        }
        // A document this ladder has not been spent on yet gets the whole thing.
        if (url != ladderUrl) {
            ladderUrl = url
            retryCount = 0
        }
        // Nothing to retry against, and no cached copy to fall back on.
        if (unreachable) {
            giveUp(view)
            return
        }
        val delayMillis = RETRY_LADDER.getOrNull(retryCount)
        if (delayMillis == null) {
            giveUp(view)
            return
        }
        retryCount++
        retryPending = true
        scheduleRetry(delayMillis) { view.loadUrl(url) }
    }

    /** The ladder is spent. */
    private fun giveUp(view: WebView) {
        view.settings.cacheMode = WebSettings.LOAD_DEFAULT
        settled = true
        binding.onLoadFailed()
    }

    /**
     * Posted to the main looper rather than through [WebView.postDelayed]: the pooled view can be
     * detached here, and a `View`'s own queue parks runnables until it is attached to a window.
     */
    private fun scheduleRetry(
        delayMillis: Long,
        navigate: () -> Unit,
    ) {
        cancelPendingRetry()
        val runnable =
            Runnable {
                pendingRetry = null
                retryPending = false
                navigate()
            }
        pendingRetry = runnable
        handler.postDelayed(runnable, delayMillis)
    }

    /** Also reached from [SharingWebViewHandle.destroy], which is not a rebind. */
    fun cancelPendingRetry() {
        pendingRetry?.let(handler::removeCallbacks)
        pendingRetry = null
    }

    override fun onRenderProcessGone(
        view: WebView,
        detail: RenderProcessGoneDetail,
    ): Boolean {
        // MUST return true — false lets the framework kill the host app, not just the sheet.
        // Reported before the `settled` guard below: the view is unusable whether or not this
        // binding still has a failure to report.
        onRendererGone()
        // A view whose renderer is gone cannot load anything, so a booked retry would only navigate
        // a corpse and keep the sheet waiting on it.
        cancelPendingRetry()
        if (!settled) {
            settled = true
            binding.onLoadFailed()
        }
        return true
    }

    private companion object {
        /**
         * Delays a main-frame failure gets before tier 3, in order. Two rungs, sized to fit inside
         * the sheet's own load budget alongside the attempts themselves — a third would expire it.
         */
        val RETRY_LADDER = listOf(300L, 900L)
    }
}
